add e2e docs (#7512)
RasaHQ/rasa: docs/docs/training-data-format.mdx

Each step can be one of the following:

  - A [user message](#user-messages), represented by **intent** and **entities**.
  - An [or statement](#or-statement), which includes two or more user messages under it
  - A bot [action](#actions)
  - A [form](#forms)
  - A [slot was set](#slots) event

>   - A [checkpoint](#checkpoints), which connects the story to another story

Author: @akelad @amn41 for my understanding: do you propose to return that to original state as well? So we present data format as if e2e doesn't exist?

Reviewer: yeah i don't think we should have the end to end format on an overview level if we still sell it as an experimental feature

Author: ok, thank you for the clarification. Reworking my changes

Reviewer 2: yes I think that's what we should do

>   - A [user message](#user-messages), represented by **intent**, **entities** or **user**.

Reviewer 3: Shouldn't it be `intent and entities` or `user`?

>   - A [user message](#user-messages), represented by **intent** and **entities** or **user**.

Reviewer 4: ```suggestion
  - A [user message](#user-messages), represented by a combination of **intent** and **entities**, or **user**.
```

Some way of clarifying which two go together, doesn't have to be phrased exactly like this

#### User Messages

> User messages are specified with the `intent:`
> key and an optional `entities:` key or with `user:` key.

Reviewer 4: ```suggestion
There are two ways to specify user messages. You can provide the `intent:`
key and an optional `entities:` key, or you can provide the `user:` key.
```

Each step can be one of the following:

>   - A [user message](#user-messages), represented by **intent** and **entities**.

Reviewer 5: ```suggestion
  - A [user message](#user-messages), represented by **intent** and **entities** or by a text.
```

Author: this is left the same intentionally, see discussion above

  - An [or statement](#or-statement), which includes two or more user messages under it.

>   - A bot [action](#actions).

Reviewer 5: ```suggestion
  - A bot [action](#actions).
  - A bot [utterance](#end-to-end-training)
```

Author: this is left the same intentionally, see discussion above

both the intent and entities (you can, however, change this behavior
using the [`use_entities`](#entities) attribute).

If you use `user:` key while writing stories, you do not have to deal with the specific
intents of the messages that are extracted by the NLU pipeline. Instead, you can simply put
the text of the user message.
We call it end-to-end training because actual text is used for predictions.

:::caution
This feature is experimental.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

:::

These end-to-end user messages follow the format:

```yaml-rasa
stories:
- story: user message structure
  steps:
    - user: the actual text of the user message  # Required
```

> In addition, you can add entity tags that will be extracted by end-to-end policy.

Reviewer 3: ```suggestion
In addition, you can add entity tags that will be extracted by the end-to-end policy.
```

The syntax for entity tags is the same as in

> [the entities in the NLU training data](./training-data-format.mdx#entities)

Reviewer 3: [the NLU training data](./training-data-format.mdx#entities)

> For example, to represent the sentence
> `I want to check my credit balance`, where `credit` is an entity:

Reviewer 3: ```suggestion
For example, the following stories contains the user utterance
`I want to check my credit balance`. By using the syntax from the NLU training data `[credit](account_type)`, we mark `credit` as an entity of type `account_type`.
```

both the intent and entities (you can, however, change this behavior
using the [`use_entities`](#entities) attribute).

If you use `user:` key while writing stories, you do not have to deal with the specific
intents of the messages that are extracted by the NLU pipeline. Instead, you can simply put
the text of the user message.

> We call it end-to-end training because actual text is used for predictions.

Reviewer 4: What I don't understand as a user is how this affects the story prediction. How is it  decided if a user message is close enough to the one ive provided to follow that story? Is it essentially "whatever this message would be classified as intent-wise based on my other NLU data (does this entry get added to training data as well?), if i get another message that falls in that intent, then this story is relevant?"

I guess this format only applies to TED so maybe that doesn't make sense, but I think we could use a bit of clarification here

Reviewer 4: Also, is this key allowed in rules, and if not, will it error as expected?

Author: you can have it rules. then it'll be string matching. I keep the idea that Rules is exactly the same as stories

Author: added explanations

> We call it end-to-end training because actual text is used for predictions.

Reviewer 5: If we call it that, then we should also add it to the glossary (are we in sync with marketing / CSE about the naming?)

Author: I don't think we have a name. It's rather name widely used in the field

:::caution
This feature is experimental.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

:::

These end-to-end user messages follow the format:

```yaml-rasa
stories:
- story: user message structure
  steps:

>     - user: the actual text of the user message  # Required

Reviewer 5: what does `# required` mean? What is it supposed to tell the user?

Author: I don't know, it was there

Author: removed it

```

In addition, you can add entity tags that will be extracted by the end-to-end policy.
The syntax for entity tags is the same as in
[the NLU training data](./training-data-format.mdx#entities)

> For example, the following stories contains the user utterance

Reviewer 3: ```suggestion
For example, the following story contains the user utterance
```

`I want to check my credit balance`. By using the syntax from the NLU training data

> `[credit](account_type)`, we mark `credit` as an entity of type `account_type`.

Reviewer 5: Not an expert on doc guidelines, but according to [this](https://www.notion.so/rasa/Tone-and-language-guidelines-cd96f21292b24c1ab7f7457c7973ec69#766266eaf5744ca094bd247409684d28) we should say `you` (#soMeta)

Author: where?

  - action: action_store_feedback
```

> 

Reviewer: pls delete

  - action: action_store_feedback
```

If you want to put bot utterance directly in the stories, use `bot:` key followed by

> the text that you want your bot to say

Reviewer 3: ```suggestion
the text that you want your bot to say.
```

:::caution
This feature is experimental.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

:::

> 

Reviewer 3: Let's add a sentence that an end-to-end story might look like this. E.g. it can be a mixture of different keys.

  - action: action_store_feedback
```

If you want to put bot utterance directly in the stories, use `bot:` key followed by

> the text that you want your bot to say.

Reviewer 4: Similarly, I have no idea how this works. Does ted just make predictions ranking "utter_greet" against "How are you?" in a greet bot and go from there? If I have some stories where I use "utter_howareyou" and others where I provide the utterance directly, is that going to confuse TED?

Author: added explanations

If you want to know more about testing head over to
[Testing Your Assistant](testing-your-assistant.mdx).

## End-to-end Training

:::caution experimental feature
End-to-end training is an experimental feature.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

:::

> With end-to-end training, you do not have to deal with the specific

Reviewer 5: ```suggestion
With [end-to-end training](stories.mdx#end-to-end-training), you do not have to deal with the specific
```

intents of the messages that are extracted by the NLU pipeline.
Instead, you can simply put the text of the user message directly in the stories,

> by using `user:` key.

Reviewer 6: For the sake of consistency, we might want to use "key" only for the actual word, not for the following colon.

```suggestion
by using `user` key.
```

These end-to-end user messages follow the format:

> ```yaml-rasa

Reviewer 6: ```suggestion
```yaml-rasa {4}
```

stories:
- story: user message structure
  steps:
    - user: the actual text of the user message
```

In addition, you can add entity tags that can be extracted
by the [TED Policy](./policies.mdx#ted-policy).
The syntax for entity tags is the same as in

> [the NLU training data](./training-data-format.mdx#entities)

Reviewer 6: ```suggestion
[the NLU training data](./training-data-format.mdx#entities).
```

For example, the following story contains the user utterance
`I want to check my credit balance`. By using the syntax from the NLU training data
`[credit](account_type)`, you can mark `credit` as an entity of type `account_type`.

```yaml-rasa {4}
stories:
- story: story with entities
  steps:
  - user: I want to check my [credit](account_type) balance
  - action: action_credit_account_balance
```

> Similarly you can put bot utterances directly in the stories,

Reviewer 6: ```suggestion
Similarly, you can put bot utterances directly in the stories,
```

> use `bot:` key followed by the text that you want your bot to say.

Reviewer 5: ```suggestion
by using the `bot:` key followed by the text that you want your bot to say.
```

> by using the `bot:` key followed by the text that you want your bot to say.

Reviewer 6: See my comment above for `user:`.

```suggestion
by using the `bot` key followed by the text that you want your bot to say.
```

> A story with only bot utterance might look like this:

Reviewer 6: ```suggestion
A story with only a bot utterance might look like this:
```

```yaml-rasa {5}
stories:
- story: story with an end-to-end response
  steps:
  - intent: greet
  - bot: Hello
```

> You can also have full end-to-end story:

Reviewer 5: How about adding a story which mixes all 4 things (`user:`, `intent`, `action`, `bot`)?

```yaml-rasa {4-5}
stories:
- story: full end-to-end story
  steps:
  - user: Hello
  - bot: Hello
```

Rasa end-to-end training is fully integrated with standard Rasa approach.

> It means that you can have mixed stories with some steps defined by actions
> and other steps defined by user or bot utterance directly.

Reviewer 6: ```suggestion
It means that you can have mixed stories with some steps defined by actions or intents
and other steps defined directly by user messages or bot responses.
```

If you want to know more about testing head over to
[Testing Your Assistant](testing-your-assistant.mdx).

## End-to-end Training

:::caution experimental feature
End-to-end training is an experimental feature.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

:::

With [end-to-end training](stories.mdx#end-to-end-training), you do not have to deal with the specific
intents of the messages that are extracted by the NLU pipeline.

> Instead, you can simply put the text of the user message directly in the stories,

Reviewer 5: ```suggestion
Instead, you can put the text of the user message directly in the stories,
```

by using `user` key.

These end-to-end user messages follow the format:

```yaml-rasa {4}
stories:
- story: user message structure
  steps:
    - user: the actual text of the user message
```

In addition, you can add entity tags that can be extracted
by the [TED Policy](./policies.mdx#ted-policy).
The syntax for entity tags is the same as in
[the NLU training data](./training-data-format.mdx#entities).
For example, the following story contains the user utterance
` I can always go for sushi`. By using the syntax from the NLU training data
`[sushi](cuisine)`, you can mark `sushi` as an entity of type `cuisine`.

```yaml-rasa {4}
stories:
- story: story with entities
  steps:
  - user: I can always go for [sushi](cuisine)
  - action: utter_suggest_cuisine
```

Similarly, you can put bot utterances directly in the stories,
by using the `bot` key followed by the text that you want your bot to say.

A story with only a bot utterance might look like this:

```yaml-rasa {7}
stories:
- story: story with an end-to-end response
  steps:
  - intent: greet
    entities:
      name: Ivan

>   - bot: Hello, Ivan!

Reviewer 7: I wonder if this is a good example? Isn't this a situation were you would always use an `utter_` action so you could output a slot value?

Author: true

Author: `Hello, a person with a name!` does it sound good 😅

```

> You can also have full end-to-end story:

Reviewer 5: ```suggestion
You can also have a full end-to-end story:
```

If you want to know more about testing head over to
[Testing Your Assistant](testing-your-assistant.mdx).

> ## End-to-end Training Stories

Reviewer: ```suggestion
## End-to-end Training
```
since we call it that on the other page, and it also fits in the sidebar better

:::caution experimental feature
End-to-end training is an experimental feature.
We introduce experimental features to get feedback from our community, so we encourage you to try it out!
However, the functionality might be changed or removed in the future.
If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).

:::

With end-to-end training, you do not have to deal with the specific
intents of the messages that are extracted by the NLU pipeline.
Instead, you can simply put the text of the user message directly in the stories,
by using `user:` key.

These end-to-end user messages follow the format:

```yaml-rasa
stories:
- story: user message structure
  steps:

>     - user: the actual text of the user message

Reviewer: should there be an action after this ?

Author: yes but in the example for intents on lines 404-410 there is no actions

Author: I just copy pasted

Reviewer 5: Ok, then how about adding an action here?

Author: done
